### PR TITLE
Actively monitor consumers - attempt to heal sick ones individually

### LINKED ISSUE #46: consumers do not recover from a network outage
I happened to be running the feed service on my laptop when the network completely went out (I pulled the plug). As you would expect, the one consumer I had started freaking out at not being able to connect to Message Hub.  However, it also did not recover once the network connection was restored.

The following messages were repeated indefinitely until I restarted the entire service:
```
WARNING:kafka.client:Node 1 connection failed -- refreshing metadata
ERROR:kafka.coordinator:Error sending GroupCoordinatorRequest_v0 to node 1 [NodeNotReadyError: 1]
WARNING:kafka.coordinator:Coordinator unknown during heartbeat -- will retry
WARNING:kafka.coordinator:Heartbeat failed ([Error 15] GroupCoordinatorNotAvailableError: ); retrying
```

### PR DESCRIPTION
**Pull request by bjustin-ibm**
Implement the "doctor" thread which monitors all consumers and attempts to heal their sickness by restarting them.

This is a pretty big change...

Consumers no longer _are_ Threads, but rather contain a ConsumerThread field. This allows the Consumer to separately monitor and replace the Thread doing the actual work. The Doctor wakes up every so often, monitors each of the Consumers and checks their health (is their actual state consistent with their desired state). If a Consumer is found to be sick, the Doctor attempts to restart that Consumer individually.

### CODE REVIEW

**Comment by dubeejw** [comment]
ERROR:root:[/.../messageHubTrigger] Uncaught exception: class State has no attribute 'Restarted'

File: provider/thedoctor.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        logging.info('[Doctor] The Doctor is in!')
+
+        while True:
+            consumers = self.consumerCollection.getCopyForRead()
+
+            for consumerId in consumers:
+                consumer = consumers[consumerId]
+                logging.debug('[Doctor] [{}] Consumer is in state: {}'.format(consumerId, consumer.currentState()))
+
+                if consumer.currentState() is Consumer.State.Dead and consumer.desiredState() is Consumer.State.Running:
```
Comment (dubeejw) [inline provider/thedoctor.py]: This seems to work fix the scrolling socket disconnect issue:
```
elif consumer.secondsSinceLastPoll() > 2 and consumer.desiredState() is Consumer.State.Running:
    logging.error('[Doctor][{}] Consumer timed-out, but should be alive!'.format(consumerId))
    consumer.restart()
```

**Comment by bjustin-ibm** [comment]
> class State has no attribute 'Restarted'

@dubeejw nice catch. This has since been fixed.

File: provider/thedoctor.py (hunk 1 above)
Comment (bjustin-ibm) [inline provider/thedoctor.py]: This makes me both happy and sad 🎭  Thanks.

Comment (bjustin-ibm) [inline provider/thedoctor.py]: 846e05b583a3aff422b504e9a6101ba9d4a7efcc

Comment (bjustin-ibm) [inline provider/thedoctor.py]: Added in 5517fbdeb569c64c6dfe28fa09d29a82ca452a6f

**Comment by bjustin-ibm** [comment]
Resolves #46, #35, #51, and #43

**Comment by bjustin-ibm** [comment]
What do you think @dubeejw? Merge it?

**Comment by dubeejw** [comment]
@bjustin-ibm, merge it.

**Pull request merged**